### PR TITLE
Latent Quantization

### PR DESCRIPTION
**Pull request by kifarid**
Added the work of @kylehkhsu https://arxiv.org/abs/2305.18378 to the repo, description and details is in the readme

### CODE REVIEW

**Comment by lucidrains** [comment]
@kifarid ah interesting

how would you tldr this paper? is it basically FSQ with a learned codebook? did you run it against FSQ?

**Comment by kifarid** [comment]
### **QLAE (Quantized Latent Autoencoder):**

- **Introduction of Latent Quantization (LQ):**
  - QLAE introduces a latent quantization (LQ) approach for autoencoders, where the latent space is quantized into discrete code vectors. This is achieved through a learnable scalar codebook per dimension, along with strong model regularization and high weight decay.

- **Efficiency Improvement Over Vector Quantization (VQ):**
  - It eliminates the need for Vector Quantization (VQ) to calculate pairwise distances between encoded vectors and codebook vectors, making it more efficient.

- **Per-Dimension Optimizable Scalar Codebook:**
  - The per-dimension optimizable scalar codebook enables the decoder to consistently assign meaning to each value.

- **Similarity to Lookup Free Quantization (LFQ):**
  - QLAE shares similarities with Lookup Free Quantization (LFQ) but is not binary. Despite not being entirely lookup-free, it still outperforms vector quantization.

- **Enhanced Disentanglement:**
  - QLAE exhibits significantly better disentanglement compared to VQ, while FSQ achieves comparable results to VQ.

*Refer to the main figure in the [latent quantization](https://arxiv.org/pdf/2305.18378.pdf) paper for a visual explanation of the inductive bias.*

image

- **Simplicity in Connection Between True Sources (c) and Data (b):**
  - The connection between true sources (c) and data (b) is the most simple and straightforward compared to any unstructured latent space, as shown in (a).

- **Discretization Process (Arrows):**
  - The discretization process (arrows) in (b) involves mapping continuous encodings (dots) to their nearest discrete codes (black circles).

- **Lockstep Movement of Codes:**
  - There is a lockstep movement of the codes due to the combinatorially defined codebook, altering the quantized representations of many datapoints (closest to this code along the specific dimension). In contrast, naively quantizing into individual vector embeddings (e) would result in codebook optimization having a comparatively local effect, affecting only datapoints with close encodings along all (less probable) dimensions.

### **Analysis:**

I am planning to conduct a more conclusive analysis of recent tokenizers, including FSQ, LFQ, and LQ. The assessment will cover generative ability (FID, precision, recall, etc.), codebook usability, disentanglement (infoMEC, introduced in the LQ paper), and additional ablation studies (e.g., comparing the effect of bounding in FSQ instead of the commitment loss). I can also share the results here once the analysis is complete.

**Comment by lucidrains** [comment]
ok sg, yea I'd like to see some head to heads to fsq and lfq first

**Comment by kylehkhsu** [comment]
@kifarid thanks for your interest in our work! I've tried FSQ in my private repo and found it to slightly underperform latent quantization in terms of disentanglement. My working conjecture is that having learnable values in LQ makes it easier for the model to shuffle things around in latent space. Very curious what your comparison of FSQ, LFQ, and LQ will yield.

**Comment by kifarid** [comment]
I have the same findings too as @kylehkhsu on MPI3D (which is relatively challenging for disentanglement). Please check the following report for the results on the disentanglement using a vanilla AE with the latent quantization repo https://api.wandb.ai/links/kifarid/og4dhjel. 
 
 I'm also currently in the process of integrating LQ with your implementation for [magvit2](https://github.com/lucidrains/magvit2-pytorch/tree/main) to provide a more comprehensive evaluation, including generative methods metrics such as FID.
 
Let me know if there are any specific tests or checks you'd like me to perform or see before accepting the pull request.

**Comment by lucidrains** [comment]
@kifarid thank you for sharing those experiments

i'll revisit this tomorrow, in the middle of another project

**Comment by lucidrains** [comment]
@kifarid if it is not too much of an ask, could you throw in an experiment for LFQ?

**Comment by kifarid** [comment]
Yeah nice! Will do, in a couple of days 🙏🏼. Thanks!

**Comment by kifarid** [comment]
@lucidrains updated the report, it seems that LFQ is lagging behind the other two, I guess probably due to the limited binary nature of it. Will try with MagVIT and let you know how it goes :)

**Comment by lucidrains** [comment]
@kifarid very interesting! thank you

**Comment by lucidrains** [comment]
@kifarid are you holding the number of codes constant across quantizers?

**Comment by kifarid** [comment]
The number of codes for the shapes3d experiment for FSQ and LQ was fixed to dim=12 and levels to 10, which corresponds to codebook size of 10^12. For LFQ, dim=22 which corresponds to 2^22 codebook size. I wanted to try for more levels to make them equal but there are some memory constraints regarding the calculation of the entropy. 

The variations of shapes3d dataset. 
| Key               | Value |
|-------------------|-------|
| value_floor_hue     | 10    |
| value_object_hue  | 10    |
| value_orientation   | 10    |
| value_scale            | 8     |
| value_shape          | 4     |
| value_wall_hue     | 15    |

So all methods can cover the variations with a good margin in the number of codes.

I also tried LQ with 10^7 and 10^8 I got a PSNR of around ~31 compared to LFQ with PSNR of around ~25.

**Comment by lucidrains** [comment]
@kifarid thanks for the update

the memory issue was brought up in another repo, but there was a solution where one can choose a subset of codes for the entropy reg. you can try setting this to `0.1` https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/lookup_free_quantization.py#L66

**Comment by kifarid** [comment]
- I will check that too, but for me it is also hard to store the codebook itself. If codebook_size is 2^32 (approximately 4.3 billion), to be able to match the 12^10 from other quantizers. the memory size of the array would be 2^32 * 8 bytes, which is approximately 34.4 gigabytes. A sequential measurement of the pairwise distance (between latents and all the codebook) is doable also for the entropy, but it will slow things a bit. 

 
- Do you think the 10^7 for LQ and the 2^22 for LFQ are comparable? If so I think the latent quantization would be a good addition to the repo. I am also getting better initial numbers for stable diffusion with latent quantization compared to the original vector quantization. 
 
- btw, regarding this part https://github.com/lucidrains/vector-quantize-pytorch/blob/3af411025c58822129e867dc6523d6a24743031f/vector_quantize_pytorch/lookup_free_quantization.py#L238 Do you think we can do better than random? I was thinking anyways since we have binary values, we can approximate the calculation by keeping the first N most significant bits fixed and only calculating the distance for the remaining bits, since the rest (with variation MSBs) have high distance anyways. What do you think?

**Comment by lucidrains** [comment]
@kifarid i see, why not bring the number of codes down from the other quantizers just for iso-code experiments?

as for your last point, sure, but that's kind of outside the scope of this PR

**Comment by lucidrains** [comment]
@kifarid hey, i'm just going to accept the PR, further research in the scalar quantization direction, despite the quibbles.

do you want to add the paper to the citation section?

**Comment by kifarid** [comment]
> @kifarid i see, why not bring the number of codes down from the other quantizers just for iso-code experiments?
> 
> as for your last point, sure, but that's kind of outside the scope of this PR

The issue is that the number of the latents in the latent quantization is specified by the number of sources or the factor of variations. 

> @kifarid hey, i'm just going to accept the PR, further research in the scalar quantization direction, despite the quibbles.
> 
> do you want to add the paper to the citation section?

There is a section in the readme for how to use LQ and also the citation should be there :). As for the research, I will do that in the upcoming days with multiple generative methods to gain a better insight. Thanks @lucidrains  :)!

**Comment by lucidrains** [comment]
@kifarid ah got it, and indeed you already had the citation in the PR; i must be blind

thank you for submitting this!
